lib/roo/base.rb

```diff
@@ -115,13 +115,11 @@ def collect_last_row_col_for_sheet(sheet)
     { first_row: first_row, first_column: first_column, last_row: last_row, last_column: last_column }
   end
 
-  %w(first_row last_row first_column last_column).each do |key|
-    class_eval <<-EOS, __FILE__, __LINE__ + 1
-      def #{key}(sheet = default_sheet)                                   # def first_row(sheet = default_sheet)
-        read_cells(sheet)                                                 #   read_cells(sheet)
-        @#{key}[sheet] ||= first_last_row_col_for_sheet(sheet)[:#{key}]   #   @first_row[sheet] ||= first_last_row_col_for_sheet(sheet)[:first_row]
-      end                                                                 # end
-    EOS
+  %i(first_row last_row first_column last_column).each do |key|
+    define_method(key) do |sheet = default_sheet|
+      read_cells(sheet)
+      instance_variable_get("@#{key}")[sheet] ||= first_last_row_col_for_sheet(sheet)[key]
```

Comment:
could you create "@#{key}" outside of the define_method block?
and when you are at it create symbol :"@#{key}" rather than string "@#{key}".

This change will avoid the creation of string "@#{key}" every time one of this method is called. It won't really make much difference since this method is not called that often, but since it can be easily avoided we should try to avoid it.

Reply:
gotcha, you're right, done. Thank you